### PR TITLE
Fix for 1512: sandbox.stub() fails on prototype props

### PR DESCRIPTION
**Pull request by fatso83**
#### Purpose (TL;DR) - mandatory
This PR restores the functionality in sandboxes that made it possible to stub methods that were higher on the prototype chain. 

#### Background 
The bug introduced in 74263dd happened as the check for a property only checked on the instance, not on the properties further up the prototype chain. That meant it was no longer possible to stub out methods on instances of a "class". See issue #1512 for background info.

#### Solution 
Simply checks for existence of the (enumerable) property.

#### How to verify - mandatory
1. Check out this branch 
2. `npm install`
3. `npm test` (will run the testcase) 
4. `git checkout HEAD~1 && npm test` should fail

### CODE REVIEW

**Comment by fatso83** [comment]
The test only checks for enumerable properties though ... I can change it to use `Object.hasOwnProperty`, but it will need to recurse up the prototype chain to the top. Do we have such a utility?

**Comment by lucasfcosta** [comment]
Hi @fatso83, excellent work, as always.

We do not have that exact utility, but we have [`get-property-descriptor`](https://github.com/sinonjs/sinon/blob/master/lib/sinon/util/core/get-property-descriptor.js), which goes up the prototype chain until it finds a property descriptor, otherwise it returns `undefined`.

I think we could use that and check if the `descriptor` returned is defined or not.

**Comment by fatso83** [comment]
Ha, I was wrong: the original code was 100% correct, and I just learned how to read the EcmaScript specification :smile: 

Some exploratory tests using `prop in object` on non-enumerable objects made me confused, as it showed that it actually could tell if a non-enumerable property was present. I thought that it was the complement of the `for ... in`, but then I 
1. found [the definition](https://www.ecma-international.org/ecma-262/5.1/#sec-11.8.7) of how the _in_ operator works, which lead me to
2. [`[[HasOperator]]`](https://www.ecma-international.org/ecma-262/5.1/#sec-8.12.6), which again led me to 
3. [`[[GetProperty]]`](https://www.ecma-international.org/ecma-262/5.1/#sec-8.12.2), which essentially is exactly what our internal util does - a recursive lookup on the prototype to find the descriptor!

So, basically I'll just merge this as it is, as it does exactly what was intended (somewhat by chance) :see_no_evil:

**Comment by mroderick** [comment]
👍

**Comment by lucasfcosta** [comment]
@fatso83 awesome!

Btw, if you're interested into doing some more research into the MetaObject Protocol I highly recommend you [this link](http://www.2ality.com/2014/12/es6-proxies.html). Which explains them and relates them to how `Proxies` work.
